File: packages/think-data-native-backend/ios/Plugin/Analysis.swift
import Foundation
import ThinkingSDK

@objc public class Analysis: NSObject {
    var  instance : ThinkingAnalyticsSDK? = nil
    @objc public func echo(_ value: String) -> String {
        print(value)
        return value
    }
    @objc func initSdk(_ appId: String?,_ url: String?)throws{
        guard let id = appId, id != nil else {
            throw NSError(domain:"appId is nil",code:500)
        }
        guard let serverUrl = url , serverUrl != nil else {
            throw NSError(domain:"serverUrl is nil",code:500)
        }
        let config = TDConfig.init(appId: id, serverUrl: serverUrl)
        instance = ThinkingAnalyticsSDK.start(with:config)
        instance?.enableAutoTrack(ThinkingAnalyticsAutoTrackEventType.eventTypeAll)
    }
    @objc func getDistinctId() -> String?{
        return instance?.getDistinctId()
    }
    @objc func setDistinctId(_ id: String?) throws {
        guard let identify = id , identify != nil else {
            throw NSError(domain:"id is nil",code:500)
        }
        if let sdk = instance {
            sdk.identify("\(identify)")
        }
    }
    @objc func signIn(_ account: String?)throws{
        if let sdk = instance {
            sdk.login("\(account!)")
        }
    }
    @objc func signOut(){
        if let sdk = instance {
            sdk.logout()
        }
    }
    @objc func track(_ eventName:String?,_ properties:[AnyHashable:Any]? = nil)throws{
        guard let event = eventName , eventName != nil else {
            throw NSError(domain:"eventName is nil",code:500)
        }
        if let sdk = instance {
            if  properties != nil && properties?.isEmpty == false {
                sdk.track(event, properties: properties)
            } else {
                sdk.track(event)
            }
        }
    }
    @objc func setPublicProperties(_ properties:[AnyHashable:Any]){
        if let sdk = instance {
            sdk.setSuperProperties(properties)
        }
    }
    @objc func unsetPublicProperties(_ property: String)throws{
        if let sdk = instance {
            sdk.unsetSuperProperty(property)
        }
    }
    @objc func clearPublicProperties()throws{
        if let sdk = instance {
            sdk.clearSuperProperties()
        }
    }
    @objc func getPublicProperties()->[AnyHashable:Any]?{
        if let sdk = instance {
            return sdk.superProperty
        } else{
            return nil
        }
    }
    @objc func eventStart(_ eventName:String?)throws{
        guard let event = eventName , event != nil else {
            throw NSError(domain:"serverUrl is nil",code:500)
        }
        if let sdk = instance {
            sdk.timeEvent(event)
        }
    }
    @objc func eventEnd(_ event:String?,_ properties:[AnyHashable : Any]? = nil)throws{
        if instance != nil {
            do{
            try track(event,properties)
            } catch {
                throw error
            }
        }
    }
    @objc func userSet(_ properties:[AnyHashable:Any]?)throws{
        guard let propertie = properties  , !propertie.isEmpty else {
            throw NSError(domain:"properties is nil",code:500)
        }
        if let sdk = instance {
            sdk.user_set(propertie)
        }
    }
    @objc func userSetOnce(_ properties:[AnyHashable:Any]?)throws{
        guard let propertie = properties , !propertie.isEmpty else {
            throw NSError(domain:"properties is nil",code:500)
        }
        if let sdk = instance {
            sdk.user_setOnce(propertie)
        }
    }
    @objc func userAdd(_ properties:[AnyHashable:Any]?)throws{
        guard let propertie = properties , propertie != nil , !propertie.isEmpty else {
            throw NSError(domain:"properties is nil",code:500)
        }
        if let sdk = instance {
            sdk.user_add(propertie)
        }
    }
    @objc func userUnSet(_ keys:[String]?){
        if let sdk = instance {
            keys?.forEach { item in
                sdk.user_unset(item)
            }
        }
    }
    
    @objc func userDelete(){
        if let sdk = instance {
            sdk.user_delete()
        }
    }
    @objc func userAppend(_ properties:[String:[Any]]?)throws{
        guard let propertie = properties , propertie != nil , !propertie.isEmpty else {
            throw NSError(domain:"properties is nil",code:500)
        }
        if let sdk = instance {
            sdk.user_append(propertie)
        }
    }
    @objc func userUniqAppend(_ properties:[String:[Any]]?)throws{
        guard let propertie = properties , propertie != nil , !propertie.isEmpty else {
            throw NSError(domain:"properties is nil",code:500)
        }
        if let sdk = instance {
            sdk.user_uniqAppend(propertie)
        }
    }
    @objc func getDeviceId() -> String {
        if let sdk = instance {
            return sdk.getDeviceId()
        }else{ return ""}
    }
}

File: packages/think-data-native-backend/android/src/main/java/com/jbcz/analysis/Analysis.kt
package com.jbcz.analysis

import android.content.Context
import android.util.Log
import cn.thinkingdata.android.BuildConfig
import cn.thinkingdata.android.TDConfig
import cn.thinkingdata.android.ThinkingAnalyticsSDK
import org.json.JSONArray
import org.json.JSONObject

class Analysis(val context: Context) {
    lateinit var sdk: ThinkingAnalyticsSDK
    fun initSdk(key: String, url: String) {

        var config = TDConfig.getInstance(context, key, url)
        ThinkingAnalyticsSDK.enableTrackLog(true)
        sdk = ThinkingAnalyticsSDK.sharedInstance(config)
        ArrayList<ThinkingAnalyticsSDK.AutoTrackEventType>().apply {
            add(ThinkingAnalyticsSDK.AutoTrackEventType.APP_INSTALL)
            add(ThinkingAnalyticsSDK.AutoTrackEventType.APP_START)
            add(ThinkingAnalyticsSDK.AutoTrackEventType.APP_END)
            add(ThinkingAnalyticsSDK.AutoTrackEventType.APP_CRASH)
            add(ThinkingAnalyticsSDK.AutoTrackEventType.APP_VIEW_SCREEN)
        }.run{
            sdk.enableAutoTrack(this)
        }
    }

    fun getDistinctId(): String {
        return sdk.distinctId
    }

    fun setDistinctId(id: String) {
        sdk.identify(id)
    }

    fun signIn(account: String) {
        sdk.login(account)
    }

    fun signOut() {
        sdk.logout()
    }

    fun track(eventName: String, map: Map<String, Any>? = null) {
        if (map == null) {
            sdk.track(eventName)
        } else {
            sdk.track(eventName, map.toJson())
        }
    }

    fun setPublicProperties(map: Map<String, Any>? = null) {
        map?.run {
            sdk.superProperties = map.toJson()
        }
    }

    fun unsetPublicProperties(key: String) {
        sdk.unsetSuperProperty(key)
    }

    fun clearPublicProperties() {
        sdk.clearSuperProperties()
    }

    fun getPublicProperties(): Map<String, Any> = HashMap<String, Any>().apply {
        var pro = sdk.superProperties
        for (iter in pro.keys()) {
            this[iter] = pro[iter]
        }
    }

    fun eventStart(eventName: String) {
        sdk.timeEvent(eventName)
    }

    fun eventEnd(eventName: String, map: Map<String, Any>? = null) {
        track(eventName, map)
    }

    fun userSet(map: Map<String, Any>? = null) {
        map?.run {
            sdk.user_set(toJson())
        }
    }

    fun userSetOnce(map: Map<String, Any>? = null) {
        map?.run {
            sdk.user_setOnce(toJson())
        }
    }

    fun userAdd(map: Map<String, Any>? = null) {
        map?.run {
            sdk.user_add(toJson())
        }
    }

    fun userUnSet(keys: Array<String>) {
        sdk.user_unset(*keys)
    }

    fun userDelete() {
        sdk.user_delete()
    }

    fun userAppend(key:String,value:Any) {
        var b=JSONObject().apply{
            put(key,value)
        }
        sdk.user_append(b)
    }

    fun userUniqAppend(key:String,value:Any) {
        var b=JSONObject().apply{
            put(key,value)
        }
        sdk.user_append(b)
    }

    fun getDeviceId(): String {
        return sdk.deviceId
    }
}